### DISCUSSION TITLE
TailwindCSS Custom Color not working, throws [plugin:vite:css] [postcss] colors is not defined

### DISCUSSION THREAD

**Discussion by justinefication** [tailwindlabs/tailwindcss#11965] [Help]
Hello, would like to ask some help on this issue. I can't seem to find anything like this on Google.

So I followed Tailwind's framework installation guide for Laravel using Vite and define my custom colors as outlined in Tailwind's Custom Color section, but whenever I run my application, it throws the following error:
>[plugin:vite:css] [postcss] colors is not defined

Here's my tailwind.config.js for your reference:
```js
/** @type {import('tailwindcss').Config} */

export default {
  content: [
    './resources/**/*.blade.php',
    './resources/**/*.js',
    './resources/**/*.vue',
  ],
  theme: {

    colors: {
        transparent: 'transparent',
        currentColor: 'currentColor',
        azure: {
            50: '#f1f7fd',
            100: '#e0edf9',
            200: '#c8dff5',
            300: '#a2ccee',
            400: '#76afe4',
            500: '#5693db',
            600: '#4279ce',
            700: '#3865bd',
            800: '#385ba9',
            900: '#2e487a',
            950: '#202d4b',
            default: '#385BA9'
        },
        tangaroa: {
            50: '#ebf8ff',
            100: '#d2f0ff',
            200: '#afe6ff',
            300: '#78d8ff',
            400: '#39c0ff',
            500: '#0c9dff',
            600: '#0078ff',
            700: '#005fff',
            800: '#004ed4',
            900: '#0347a5',
            950: '#041735',
            default: '#041735'
        },
        rose: {
            50: '#fff1f2',
            100: '#ffe4e6',
            200: '#fecdd3',
            300: '#fda4af',
            400: '#fb7185',
            500: '#f43f5e',
            600: '#e11d48',
            700: '#be123c',
            800: '#9f1239',
            900: '#881337',
            950: '#4c0519',
            default: '#e11d48'
        },
        amber: {
            50: '#fffbeb',
            100: '#fef3c7',
            200: '#fde68a',
            300: '#fcd34d',
            400: '#fbbf24',
            500: '#f59e0b',
            600: '#d97706',
            700: '#b45309',
            800: '#92400e',
            900: '#78350f',
            950: '#451a03',
            default: '#fbbf24'
        },
        gray: {
            50: '#f9fafb',
            100: '#f3f4f6',
            200: '#e5e7eb',
            300: '#d1d5db',
            400: '#9ca3af',
            500: '#6b7280',
            600: '#4b5563',
            700: '#374151',
            800: '#1f2937',
            900: '#111827',
            950: '#030712',
        },
        primary: colors.azure,
        secondary: colors.tangaroa,
        danger: colors.rose,
        warning: colors.amber,
        neutral: colors.gray
    },
    extend: {},
  },
  plugins: [],
}
```

**Comment by wongjn** [comment]
It seems like you are referencing a `colors` variable:

```js
primary: colors.azure,
secondary: colors.tangaroa,
danger: colors.rose,
warning: colors.amber,
neutral: colors.gray
```

But this `colors` variable has not been defined in file, thus the error.

**Reply by justinefication** [reply]
Hi @wongjn, thanks for the response. Those are color alias and it is how they're described in the tailwind docs, but when I remove those aliases the application runs, but nothing happens when I use color classes, e.g.: `text-azure`

**Reply by wongjn** [reply]
In [the Tailwind documentation](https://tailwindcss.com/docs/customizing-colors#using-the-default-colors), the `colors` variable is defined before use. For example:

```js
const colors = require('tailwindcss/colors')

module.exports = {
  theme: {
    colors: {
      transparent: 'transparent',
      current: 'currentColor',
      black: colors.black,
      white: colors.white,
      gray: colors.gray,
      emerald: colors.emerald,
      indigo: colors.indigo,
      yellow: colors.yellow,
    },
  },
}
```

`colors` is defined in the first line:

```js
const colors = require('tailwindcss/colors')
```

Your Tailwind configuration has no such `color` variable definition before the use of the `colors` variable, thus the error.

Additionally, there is no `azure` color that would be available from [Tailwind's color pallette](https://github.com/tailwindlabs/tailwindcss/blob/v3.3.3/src/public/colors.js) (or `tangaroa` for that matter).

**Reply by justinefication** [reply]
The above snippet was included in using the default color library used by Tailwind, specifically using selectively the colors from the Tailwind provided color library.

Also, I tried including the snippet

```js

const colors = require('tailwindcss/colors');
```

But using color classes such as `text-azure`, azure being defined above does not change the color of the text.

**Reply by justinefication** [reply]
> Additionally, there is no azure color that would be available from Tailwind's color pallette (or tangaroa for that matter).

I'm defining custom colors.

**Reply by wongjn** [reply]
> The above snippet was included in using the default color library used by Tailwind, specifically using selectively the colors from the Tailwind provided color library.

Which above snippet? There are many snippets above!

> But using color classes such as `text-azure`, azure being defined above does not change the color of the text.

Consider checking:

- Tailwind has compiled since you added the class.
- The `content` file globs sufficiently cover the file that the `text-azure` class name appears in.
- [String concatenation/interpolation is not used to build the class name](https://tailwindcss.com/docs/content-configuration#dynamic-class-names). 

> > Additionally, there is no `azure` color that would be available from [Tailwind's color pallette](https://github.com/tailwindlabs/tailwindcss/blob/v3.3.3/src/public/colors.js) (or `tangaroa` for that matter).
>
> I'm defining custom colors.

This point was in reference to

```js
primary: colors.azure,
secondary: colors.tangaroa,
```

Assuming `colors` is from `require('tailwindcss/colors')`, `azure` and `tangaroa` would not exist from this `require`d variable.

**Reply by justinefication** [reply]
> Which above snippet? There are many snippets above!

**Let's remain to be civilized shall we?**

The following snippets you replied, came from Tailwind's [Using the default colors](https://tailwindcss.com/docs/customizing-colors#using-the-default-colors) section of documentation

```js
const colors = require('tailwindcss/colors')

module.exports = {
  theme: {
    colors: {
      transparent: 'transparent',
      current: 'currentColor',
      black: colors.black,
      white: colors.white,
      gray: colors.gray,
      emerald: colors.emerald,
      indigo: colors.indigo,
      yellow: colors.yellow,
    },
  },
}
```
Where the documentation states and I quote:
> **If you don’t have a set of completely custom colors** in mind for your project, you can curate your **colors from our default palette by importing** tailwindcss/colors in your configuration file

The documentation is very specific that `if you don't have a set of complete colors`. In contrary, that's not what I'm trying to accomplish. What I'm trying to accomplish is to define my custom set of colors which is described and documented in [Using custom colors](https://tailwindcss.com/docs/customizing-colors#using-custom-colors). Where it says and I quote:
> If you’d like to completely replace the default color palette with your own custom colors, add your colors directly under the theme.colors section of your configuration file

And goes on and provide the following example code snippet, **beginning from the very top of the file**
```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  theme: {
    colors: {
      transparent: 'transparent',
      current: 'currentColor',
      'white': '#ffffff',
      'purple': '#3f3cbb',
      'midnight': '#121063',
      'metal': '#565584',
      'tahiti': '#3ab7bf',
      'silver': '#ecebff',
      'bubble-gum': '#ff77e9',
      'bermuda': '#78dcca',
    },
  },
}
```

The code snippet clearly doesn't include a `const colors = require('tailwindcss/colors')` code in there, is there?

Furthermore, [Aliasing color names](https://tailwindcss.com/docs/customizing-colors#aliasing-color-names) declare the alias inside of the `colors` object under the `theme` object, consistent with how I implemented my original code snippet, am I correct?

> Assuming colors is from require('tailwindcss/colors'), azure and tangaroa would not exist from this required variable.

How can we assume if I clearly stated in my question the following statement?
> . . . and define my custom colors as outlined in Tailwind's Custom Color section . . .

**Reply by wongjn** [reply]
> The code snippet clearly doesn't include a `const colors = require('tailwindcss/colors')` code in there, is there?

Indeed, because it does not use a `colors` variable in the content of the file. This is the major difference in your file, since you *do* reference a `colors` variable:

```js
primary: colors.azure,
secondary: colors.tangaroa,
danger: colors.rose,
warning: colors.amber,
neutral: colors.gray
```

But this `colors` variable has not been defined in file, thus why you were getting the error:

```
[plugin:vite:css] [postcss] colors is not defined
```
So even if you did not want `colors` from `require('tailwindcss/colors')`, you'd still need some kind of `colors` variable declared or removal of the references to this `colors` variable.

---

> Furthermore, [Aliasing color names](https://tailwindcss.com/docs/customizing-colors#aliasing-color-names) declare the alias inside of the colors object under the theme object, consistent with how I implemented my original code snippet, am I correct?

No, you are not correct. Your original code snippet is inconsistent since the example code snippet provided at [Aliasing color names](https://tailwindcss.com/docs/customizing-colors#aliasing-color-names) has

```js
const colors = require('tailwindcss/colors')
```

As the first line. This line is missing from your configuration file, thus the `colors` variable has not been defined in file, thus why you were getting the error:

```
[plugin:vite:css] [postcss] colors is not defined
```

**Reply by Zehad-10** [reply]
You need to use your color `text-azure-50` not `text-azure`

**Reply by wongjn** [reply]
Ah yes, great find! That or use capital `DEFAULT` if you want `text-azure` to be a valid class.
